[WPE] Click count leaking through successive layout test within the same WPEView

lauromoura opened the pull request:
#### 73636813c28fde71e3b86b395acabaa71e2328a1
<pre>
[WPE] Click count leaking through successive layout test within the same WPEView
<a href="https://bugs.webkit.org/show_bug.cgi?id=295252">https://bugs.webkit.org/show_bug.cgi?id=295252</a>

Reviewed by NOBODY (OOPS!).

Add a new function to reset click count tracking between page loads, avoiding
accidental double and triple clicks.

Found this issue while rebaselining the fast/html/details-* tests after
296203@main. Some rebaselines were flaky as the tests were affected by
this when running together.

This helps WPE mirror the GTK4 WebView behavior of resetting the click
count between page loads.

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::didCommitLoadForMainFrame): Reset click count
between page loads.
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_clear_press_count): Added.
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73636813c28fde71e3b86b395acabaa71e2328a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30424 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20856 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116790 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61029 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112724 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31104 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39010 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84219 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113709 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24846 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99737 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24214 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17870 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60583 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94239 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17929 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119580 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37806 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28100 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93181 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38182 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96000 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93005 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38038 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15787 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33780 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37699 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43169 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37360 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40698 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39067 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -996,6 +996,20 @@ guint wpe_view_compute_press_count(WPEView* view, gdouble x, gdouble y, guint bu
+ * wpe_view_clear_press_count
+ * @view: a #WPEView
+ *
+ * Clear any button press being tracket.
```
Typo: tracket.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.h:
```diff
@@ -138,6 +138,7 @@ WPE_API guint                   wpe_view_compute_press_count           (WPEView
                                                                         gdouble             y,
                                                                         guint               button,
                                                                         guint32             time);
+WPE_API void                    wpe_view_clear_press_count             (WPEView            *view);
```
It seems counter-intuitive that an external entity has to tell the view that it has to clear something, because of a navigation/load action. I guess there is never a situation where we want to preserve that press count between such events, no?

lauromoura replied:
IIUC, `WPEView` is in charge of lower level operations like delivering events to the `WebKitWebView` and forwarding buffers received from it. Something like a "shell" around `WebKitWebView`. So, `WPEView` would not have knowledge of web operations like loading, navigation, etc.

Maybe instead of the `EventSender` directly calling this new function, should we make `WebKitWebView` call it when those cases happen, like loading a new page? 

Another option, if we want to avoid creating a new function, would be creating something like a `WPEEventType::WPE_EVENT_RESET`, that, when sent through `wpe_view_event(event)`, tells the `WPEView` to reset its event state tracking.

carlosgcampos replied:
That's not GTK4 but GTK3 and it's not used by layout tests. EventSenderProxy always passes its computed click count value to the synthesized event that it's used directly instead of the value computed by the web view in the case of GTK3 or the value computed by GTK in the case of GTK4. The reason why this fails in WPE is becase we use  WPEView to compute the value, which is the same for tests run by the same WTR. All other ports compute the click count in EventSenderProxy. A new EventSenderProxy is created for test even in the same WTR so click count is always reset for every test. So, I think we should follow all other ports and compute the click counter in EventSenderProxy instead of using WPEView.

carlosgcampos replied:
See https://github.com/WebKit/WebKit/pull/47813

lauromoura replied:
Ouch, my mistake regarding GTK3/4, sorry. Tested #47813 and the tests are working fine with it. I'll close this one. Thanks!

lauromoura committed:
[WPE] Click count leaking through successive layout test within the same WPEView
https://bugs.webkit.org/show_bug.cgi?id=295252

Reviewed by NOBODY (OOPS!).

Add a new function to reset click count tracking between page loads, avoiding
accidental double and triple clicks.

Found this issue while rebaselining the fast/html/details-* tests after
296203@main. Some rebaselines were flaky as the tests were affected by
this when running together.

This helps WPE mirror the GTK4 WebView behavior of resetting the click
count between page loads.

* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::didCommitLoadForMainFrame): Reset click count
between page loads.
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_clear_press_count): Added.
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
